File: python/src/solutions/5kyu/paginating_a_huge_book/solution_paginating_a_huge_book.py
"""https://www.codewars.com/kata/55905b7597175ffc1a00005a"""

from math import floor, log10


def page_digits(pages):
    result = 0
    digits = floor(log10(pages)) + 1

    for n in range(1, digits):
        result += n * (10 ** n - 10 ** (n - 1))

    result += (pages - (10 ** (digits - 1) - 1)) * digits

    return result

File: c/src/solutions/5kyu/paginating_a_huge_book/solution_paginating_a_huge_book.c
/*
 * https://www.codewars.com/kata/55905b7597175ffc1a00005a
 */

#include "solution_paginating_a_huge_book.h"

#define ull unsigned long long

ull page_digits(ull pages) {
    ull start = 0, place = 10, result = 0;

    while (pages > start) {
        result += pages - start;
        start = place - 1;
        place *= 10;
    }

    return result;
}
